quickjs.c

```diff
@@ -54755,8 +54765,12 @@ JSValue JS_NewUint8ArrayCopy(JSContext *ctx, const uint8_t *buf, size_t len)
     return js_new_uint8array(ctx, buffer);
 }
 
-JS_BOOL JS_IsUint8Array(JSValue obj) {
-    return JS_GetClassID(obj) == JS_CLASS_UINT8_ARRAY;
+int JS_GetTypedArrayType(JSValue obj)
+{
+    JSClassID class_id = JS_GetClassID(obj);
+    int mask = -((class_id >= JS_CLASS_UINT8C_ARRAY) & (class_id <= JS_CLASS_FLOAT64_ARRAY));
+    int offset = (class_id - JS_CLASS_UINT8C_ARRAY) & mask;
+    return offset | (~mask & -1);
```

Comment:
I suppose you took this approach to ensure branchless code generation... But reading the code and trying to make sure it works as intended gives me a headache.  This breaks a golden rule: write the code as simply as possible.

I would suggest this simple code instead and let the compiler generate optimal code:
```
if (class_id >= JS_CLASS_UINT8C_ARRAY && class_id <= JS_CLASS_FLOAT64_ARRAY)
    return class_id - JS_CLASS_UINT8C_ARRAY;
else
    return -1;
```

Both *gcc* and *clang* generate branchless code for the above:

https://godbolt.org/z/hvhrnzMsx

---
Kernighan's Law - Debugging is twice as hard as writing the code in the first place. Therefore, if you write the code as cleverly as possible, you are, by definition, not smart enough to debug it.

Reply:
👍 good observation.